rescan_test.go

```diff
@@ -0,0 +1,680 @@
+// retry blocks under different scenarios.
+func TestRescanRetryBlocks(t *testing.T) {
+	t.Parallel()
+
+	// Define some helper closures to ease the ability of marking a block as
+	// a retry block. This is done by replacing the GetCFilter
+	// implementation with one that returns a nil filter, causing the Rescan
+	// to retry fetching it.
+	filterQueried := make(chan chainhash.Hash, 1)
+	filterNotFound := func(hash chainhash.Hash,
+		_ wire.FilterType, _ ...QueryOption) (*gcs.Filter, error) {
+
+		filterQueried <- hash
+		return nil, nil
+	}
+
+	assertFilterQueried := func(t *testing.T, hash chainhash.Hash) {
+		t.Helper()
+
+		var hashQueried chainhash.Hash
+		select {
+		case hashQueried = <-filterQueried:
+			if !hashQueried.IsEqual(&hash) {
+				t.Fatalf("expected to query filter %v, got %v",
+					hash, hashQueried)
+			}
+		case <-time.After(time.Second):
+			t.Fatal("expected rescan to query for filter")
+		}
+	}
+
+	testCases := []struct {
+		name string
+		run  func(ctx *rescanTestContext)
+	}{
+		{
+			// We should not attempt to process new blocks if we're
+			// retrying a previous one.
+			name: "new block after retry",
+			run: func(ctx *rescanTestContext) {
+				// Override the GetCFilter implementation to
+				// mark the following blocks as retries.
+				ctx.chain.overrideGetCFilter(filterNotFound)
+
+				// Adding new blocks to the chain should result
+				// in the first retry block being queried until
+				// successful.
+				block1 := ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block1.Hash)
+				block2 := ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block1.Hash)
+
+				// Revert back to the original GetCFilter
+				// implementation, causing us to properly notify
+				// the blocks in order.
+				ctx.chain.overrideGetCFilter(nil)
+
+				ctx.recvBlockConnected(block1)
+				ctx.recvBlockConnected(block2)
+			},
+		},
+		{
+			// We should stop retrying blocks for which we've
+			// received disconnected notifications for.
+			name: "reorg n-1 retry blocks",
+			run: func(ctx *rescanTestContext) {
+				// Override the GetCFilter implementation to
+				// mark the following blocks as retries.
+				ctx.chain.overrideGetCFilter(filterNotFound)
+
+				// Adding new blocks to the chain should result
+				// in the first retry block being queried until
+				// successful.
+				block := ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block.Hash)
+				ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block.Hash)
+				ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block.Hash)
+
+				// We'll then reorg back to the first retry
+				// block. This should cause us to stop retrying
+				// the reorged blocks.
+				ctx.chain.rollbackToHeight(block.Height, true)
+
+				// We should then receive a notification for it
+				// once we revert back to the original
+				// GetCFilter implementation.
+				ctx.chain.overrideGetCFilter(nil)
+				ctx.recvBlockConnected(block)
```

Comment:
Doesn't this need more assertions to check that the right queries are executed now? Or is that all implied by the block connected ntfn?

Reply:
We rely on the notification of the block being delivered if its filter is fetched. Added the assertion anyway though since we now can.